deprecate PairwiseParallel from test

tianyu-l opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114314

**Summary**
To solve issue #113706:
1. replace `PariwiseParallel` with `ColwiseParallel` and `RowwiseParallel`.
2. replace the input of ColwiseParallel from `make_input_replicate_1d` and `make_output_replicate_1d` to `input_layouts` and `output_layouts`.
3. deprecate the tests for `_parallelize_mlp` as it only supports `PariwiseParallel`.

**Test Plan**
`pytest pytorch/test/distributed/tensor/parallel/`


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225 @kiukchung @d4l3k @lucasllc

tianyu-l committed:
deprecate PairwiseParallel from test

[ghstack-poisoned]

tianyu-l commented:
/easycla

tianyu-l commented:
issue https://github.com/pytorch/pytorch/issues/113706

wanchaol reviewed (approved):
change lgtm, thanks!

XilunWu reviewed (approved):
Congrats on your first PyTorch PR!!

tianyu-l commented:
/easycla

wanchaol commented:
@tianyu-l just to clarify, there's going to be a follow up PR to remove `SequenceParallel` style in tests right?

wanchaol commented:
Actually there're also bulk of callsites that are using `PairwiseParallel` outside of `distributed/tensor/parallel` folder, i.e. https://github.com/pytorch/pytorch/blob/main/test/distributed/checkpoint/test_tp_checkpoint.py#L15

Feel free to change them either in this PR or have follow up PRs to remove them

tianyu-l commented:
> @tianyu-l just to clarify, there's going to be a follow up PR to remove `SequenceParallel` style in tests right?

There doesn't seem to be tests for SequenceParallel in `distributed/tensor/parallel`, only one [here](https://github.com/pytorch/pytorch/blob/d5d62e85615fdf345e0556a9d8edbee2d3c64ae2/test/distributed/test_fake_pg.py#L20). Shall I replace that as well?

wanchaol commented:
> > @tianyu-l just to clarify, there's going to be a follow up PR to remove `SequenceParallel` style in tests right?
> 
> There doesn't seem to be tests for SequenceParallel in `distributed/tensor/parallel`, only one [here](https://github.com/pytorch/pytorch/blob/d5d62e85615fdf345e0556a9d8edbee2d3c64ae2/test/distributed/test_fake_pg.py#L20). Shall I replace that as well?

I see, saw another test in [here](https://github.com/pytorch/pytorch/blob/main/test/distributed/fsdp/test_fsdp_tp_integration.py#L252) Feel free to replace them in this PR or in a follow up PR :)

XilunWu commented:
> > > @tianyu-l just to clarify, there's going to be a follow up PR to remove `SequenceParallel` style in tests right?
> > 
> > 
> > There doesn't seem to be tests for SequenceParallel in `distributed/tensor/parallel`, only one [here](https://github.com/pytorch/pytorch/blob/d5d62e85615fdf345e0556a9d8edbee2d3c64ae2/test/distributed/test_fake_pg.py#L20). Shall I replace that as well?
> 
> I see, saw another test in [here](https://github.com/pytorch/pytorch/blob/main/test/distributed/fsdp/test_fsdp_tp_integration.py#L252) Feel free to replace them in this PR or in a follow up PR :)

@tianyu-l I suggest we merge this PR first because we already have the trunk CI tests pass and create another PR to address Wanchao's comment for faster development iterations.

tianyu-l commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but this seems to fail periodic multigpu tests' -c nosignal

@wanchaol and @awgu discovers the failures in https://github.com/pytorch/pytorch/pull/114390#issuecomment-1823721757

huydhn commented:
I have added `ciflow/periodic` to the PR to run the test here, please help rebase to main after https://github.com/pytorch/pytorch/pull/114390 lands because it's also trying to fix another failure on the same job.

awgu commented:
Hi @tianyu-l. Sorry for reverting your PR. I found locally if I patch this, I can reproduce the failing unit test:
```
python test/distributed/tensor/parallel/test_fsdp_2d_parallel.py -k test_2d_e2e_training_not_use_orig_params
```

```
======================================================================
ERROR: test_2d_e2e_training_not_use_orig_params (__main__.TestNew2dParallelTraining)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/andgu/pytorch/torch/testing/_internal/common_distributed.py", line 533, in wrapper
    self._join_processes(fn)
  File "/data/users/andgu/pytorch/torch/testing/_internal/common_distributed.py", line 752, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/data/users/andgu/pytorch/torch/testing/_internal/common_distributed.py", line 802, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 1 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/data/users/andgu/pytorch/torch/testing/_internal/common_distributed.py", line 649, in run_test
    getattr(self, test_name)()
  File "/data/users/andgu/pytorch/torch/testing/_internal/common_distributed.py", line 535, in wrapper
    fn()
  File "/data/users/andgu/pytorch/torch/testing/_internal/common_utils.py", line 2652, in wrapper
    method(*args, **kwargs)
  File "/data/users/andgu/pytorch/torch/testing/_internal/distributed/_tensor/common_dtensor.py", line 193, in wrapper
    func(self, *args, **kwargs)  # type: ignore[misc]
  File "/data/users/andgu/pytorch/torch/testing/_internal/common_distributed.py", line 174, in wrapper
    return func(*args, **kwargs)
  File "/data/users/andgu/pytorch/test/distributed/tensor/parallel/test_fsdp_2d_parallel.py", line 202, in test_2d_e2e_training_not_use_orig_params
    self._test_2d_e2e_training(recompute_activation=True)
  File "/data/users/andgu/pytorch/test/distributed/tensor/parallel/test_fsdp_2d_parallel.py", line 179, in _test_2d_e2e_training
    self.assertEqual(output, output_2d)
  File "/data/users/andgu/pytorch/torch/testing/_internal/common_utils.py", line 3576, in assertEqual
    raise error_metas.pop()[0].to_error(
AssertionError: Tensor-likes are not close!

Mismatched elements: 28 / 48 (58.3%)
Greatest absolute difference: 0.06779798865318298 at index (1, 2) (up to 1e-05 allowed)
Greatest relative difference: 0.906391978263855 at index (1, 4) (up to 1.3e-06 allowed)
```

wanchaol commented:
Thanks for figuring out the cause @awgu!

@tianyu-l Let's remove the recompute_activation if branch and reland this PR. I think we will need to revisit the `input_reshard` API about why it failed. Since this API was not widely used, we can disable the test for it for now

tianyu-l committed:
Update on "deprecate PairwiseParallel from test"


**Summary**
To solve issue #113706:
1. replace `PariwiseParallel` with `ColwiseParallel` and `RowwiseParallel`.
2. replace the input of ColwiseParallel from `make_input_replicate_1d` and `make_output_replicate_1d` to `input_layouts` and `output_layouts`.
3. deprecate the tests for `_parallelize_mlp` as it only supports `PariwiseParallel`.

**Test Plan**
`pytest pytorch/test/distributed/tensor/parallel/`


[ghstack-poisoned]

tianyu-l commented:
disabled `input_reshard` to pass tests

tianyu-l committed:
Update on "deprecate PairwiseParallel from test"


**Summary**
To solve issue #113706:
1. replace `PariwiseParallel` with `ColwiseParallel` and `RowwiseParallel`.
2. replace the input of ColwiseParallel from `make_input_replicate_1d` and `make_output_replicate_1d` to `input_layouts` and `output_layouts`.
3. deprecate the tests for `_parallelize_mlp` as it only supports `PariwiseParallel`.

**Test Plan**
`pytest pytorch/test/distributed/tensor/parallel/`


[ghstack-poisoned]

tianyu-l commented:
@pytorchbot rebase -b main

tianyu-l committed:
Update on "deprecate PairwiseParallel from test"


**Summary**
To solve issue #113706:
1. replace `PariwiseParallel` with `ColwiseParallel` and `RowwiseParallel`.
2. replace the input of ColwiseParallel from `make_input_replicate_1d` and `make_output_replicate_1d` to `input_layouts` and `output_layouts`.
3. deprecate the tests for `_parallelize_mlp` as it only supports `PariwiseParallel`.

**Test Plan**
`pytest pytorch/test/distributed/tensor/parallel/`


[ghstack-poisoned]

wanchaol commented on test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:
```diff
@@ -246,9 +249,11 @@ def test_2d_state_dict(self, is_even_sharded_model):
-        model_2d = parallelize_module(
-            simple_model().cuda(), tp_mesh, PairwiseParallel()
-        )
+        parallelize_plan = {
```
looks like there's a legit CI failure that's related to `SimpleModelUneven` where in the `SimpleModelUneven`  case it defined nn.Sequential instead of nn.Module, I think we should change the `SimpleModelUneven`  to define the same as `SimpleModel` instead of using `nn.Sequential` as we just want to test the Linear size to be unevenly divided, not testing nn.Sequential

tianyu-l replied:
@wz337 Iris told me the tests would still fail even if we make such changes, because of potential size mismatch. I cannot verify locally because of devgpu issues.

Per Iris's suggestion, I'm disabling all three tests involving the sequential SimpleModelUneven for now to see if tests from ciflow/periodic can pass so that I can land. Meanwhile I'll be working on fixing devgpu.

fduwjj replied:
Yes, for `SimpleModelUneven` it is an no-op for currently TP API if using `PairwiseParallel()` that's why test passed previously.

tianyu-l committed:
Update on "deprecate PairwiseParallel from test"


**Summary**
To solve issue #113706:
1. replace `PariwiseParallel` with `ColwiseParallel` and `RowwiseParallel`.
2. replace the input of ColwiseParallel from `make_input_replicate_1d` and `make_output_replicate_1d` to `input_layouts` and `output_layouts`.
3. deprecate the tests for `_parallelize_mlp` as it only supports `PariwiseParallel`.

**Test Plan**
`pytest pytorch/test/distributed/tensor/parallel/`


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu

[ghstack-poisoned]

fduwjj commented on test/distributed/tensor/parallel/test_parallelize_api.py:
```diff
@@ -217,40 +195,15 @@ def test_parallelize_mlp_with_module_api_nested(self):
             {
                 "dummy_encoder.net1": ColwiseParallel(
-                    make_input_replicate_1d, make_output_replicate_1d
+                    input_layouts=Replicate(), output_layouts=Replicate()
```
For `ColwiseParallel`, input_layouts by default is `Replicate()`, https://github.com/pytorch/pytorch/blob/11277cc51008054a13c455d6ddfb12a333b54aa4/torch/distributed/tensor/parallel/style.py#L455. So you can just set output_layouts here. We changed the default value of `_prepare_input` several times in the past, that's why we set it in the test previously.

fduwjj commented on test/distributed/tensor/parallel/test_tp_examples.py:
```diff
@@ -62,23 +56,15 @@ def _test_mlp_training_e2e(self, is_seq_parallel=False, recompute_activation=Fal
             else RowwiseParallel(),
         }
-        model_tp = parallelize_module(model_tp, device_mesh, parallel_style)
-        if recompute_activation:
-            model_tp = input_reshard(
-                checkpoint_wrapper(
-                    model_tp, checkpoint_impl=CheckpointImpl.NO_REENTRANT
-                ),
-                device_mesh,
-                None if is_seq_parallel else 0,
-            )
```
why do we remove this?

fduwjj reviewed (changes requested):
Edit: Opps, didn't see the long string of comments. I have one comment originally about `recompute_activations`. Instead of removing it completely, shall we only disable it in the test for `not_use_orig_params`?

tianyu-l commented:
@pytorchbot rebase -b main

tianyu-l committed:
Update on "deprecate PairwiseParallel from test"


**Summary**
To solve issue #113706:
1. replace `PariwiseParallel` with `ColwiseParallel` and `RowwiseParallel`.
2. replace the input of ColwiseParallel from `make_input_replicate_1d` and `make_output_replicate_1d` to `input_layouts` and `output_layouts`.
3. deprecate the tests for `_parallelize_mlp` as it only supports `PariwiseParallel`.

**Test Plan**
`pytest pytorch/test/distributed/tensor/parallel/`


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu

[ghstack-poisoned]

tianyu-l committed:
Update on "deprecate PairwiseParallel from test"


**Summary**
To solve issue #113706:
1. replace `PariwiseParallel` with `ColwiseParallel` and `RowwiseParallel`.
2. replace the input of ColwiseParallel from `make_input_replicate_1d` and `make_output_replicate_1d` to `input_layouts` and `output_layouts`.
3. deprecate the tests for `_parallelize_mlp` as it only supports `PariwiseParallel`.

**Test Plan**
`pytest pytorch/test/distributed/tensor/parallel/`


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu

[ghstack-poisoned]

wanchaol commented on test/distributed/tensor/parallel/test_tp_examples.py:
```diff
@@ -122,7 +135,9 @@ def _test_mlp_inference(self, device_mesh):
     @with_comms
     @parametrize("is_seq_parallel", [True, False])
+    # TODO: need to revisit input_reshard API about why it failed multi-gpu tests.
     @parametrize("recompute_activation", [True, False])
```
did we disable the `recompute_activation` option?

tianyu-l replied:
It doesn't fail the test here. It only breaks the test in `test_fsdp_2d_parallel.py`. So I kept it here. Shall we disable it altogether?

tianyu-l replied:
disabled altogether

tianyu-l committed:
Update on "deprecate PairwiseParallel from test"


**Summary**
To solve issue #113706:
1. replace `PariwiseParallel` with `ColwiseParallel` and `RowwiseParallel`.
2. replace the input of ColwiseParallel from `make_input_replicate_1d` and `make_output_replicate_1d` to `input_layouts` and `output_layouts`.
3. deprecate the tests for `_parallelize_mlp` as it only supports `PariwiseParallel`.

**Test Plan**
`pytest pytorch/test/distributed/tensor/parallel/`


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu

[ghstack-poisoned]

tianyu-l commented:
> Edit: Opps, didn't see the long string of comments. I have one comment originally about `recompute_activations`. Instead of removing it completely, shall we only disable it in the test for `not_use_orig_params`?

changed, thanks!

tianyu-l committed:
Update on "deprecate PairwiseParallel from test"


**Summary**
To solve issue #113706:
1. replace `PariwiseParallel` with `ColwiseParallel` and `RowwiseParallel`.
2. replace the input of ColwiseParallel from `make_input_replicate_1d` and `make_output_replicate_1d` to `input_layouts` and `output_layouts`.
3. deprecate the tests for `_parallelize_mlp` as it only supports `PariwiseParallel`.

**Test Plan**
`pytest pytorch/test/distributed/tensor/parallel/`


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu

[ghstack-poisoned]

tianyu-l commented:
@pytorchbot rebase -b viable/strict

tianyu-l committed:
Update on "deprecate PairwiseParallel from test"


**Summary**
To solve issue #113706:
1. replace `PariwiseParallel` with `ColwiseParallel` and `RowwiseParallel`.
2. replace the input of ColwiseParallel from `make_input_replicate_1d` and `make_output_replicate_1d` to `input_layouts` and `output_layouts`.
3. deprecate the tests for `_parallelize_mlp` as it only supports `PariwiseParallel`.

**Test Plan**
`pytest pytorch/test/distributed/tensor/parallel/`


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu

[ghstack-poisoned]

tianyu-l committed:
Update on "deprecate PairwiseParallel from test"


**Summary**
To solve issue #113706:
1. replace `PariwiseParallel` with `ColwiseParallel` and `RowwiseParallel`.
2. replace the input of ColwiseParallel from `make_input_replicate_1d` and `make_output_replicate_1d` to `input_layouts` and `output_layouts`.
3. deprecate the tests for `_parallelize_mlp` as it only supports `PariwiseParallel`.

**Test Plan**
`pytest pytorch/test/distributed/tensor/parallel/`


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu

[ghstack-poisoned]

tianyu-l commented:
@pytorchbot merge
